### PR TITLE
boards/stm32f4: Add DMA config for SPI

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds valid DMA configuration for the SPI peripherals of the boards containing an STM32F4 microcontroller. This adds the DMA as a provided feature for these boards and adds config to make it useable, but doesn't enable the DMA by default for SPI.

### Testing procedure

*with* `FEATURES_REQUIRED += periph_dma`, testing is to be done with `tests/periph_spi`. For a thorough test, check if the bytes send and received match what is expected. Otherwise simply triggering a bidirectional transaction should be sufficient, an incorrect DMA config will not set up the right interrupts and channels and will cause the SPI transfer to hang infinitely:

- [ ] msbiot
- [x] nucleo-f401re
- [ ] nucleo-f410rb
- [ ] nucleo-f411re
- [ ] nucleo-f412zg
- [ ] nucleo-f413zh
- [ ] nucleo-f429zi
- [x] nucleo-f446re
- [ ] nucleo-f446ze
- [ ] pyboard
- [x] stm32f429i-disc1
- [ ] stm32f4discovery
- [ ] ublox-c030-u201

Also please run the very short `tests/periph_dma` to verify that the UART shell doesn't break:

- [ ] msbiot
- [x] nucleo-f401re
- [ ] nucleo-f410rb
- [ ] nucleo-f411re
- [ ] nucleo-f412zg
- [ ] nucleo-f413zh
- [ ] nucleo-f429zi
- [x] nucleo-f446re
- [ ] nucleo-f446ze
- [ ] pyboard
- [x] stm32f429i-disc1
- [ ] stm32f4discovery
- [ ] ublox-c030-u201

### Issues/PRs references

Depends on #14090 to not break the shell (uart) completely when enabling DMA.

### CODE REVIEW

**Comment by bergzand** [comment]
No longer waiting for other PRs! Rebased.

**Comment by aabadie** [comment]
It will be difficult to test all boards but I can start by comparing with the datasheets.

**Comment by bergzand** [comment]
I've tried to reuse the config between boards as much as possible and use DMA streams/channels available on all MCU versions.

Also let me know if it is easier to split this PR into multiple separate PRs :)

**Comment by aabadie** [comment]
I just tested `tests/pkg_lvgl` on stm32f429disc1 and it's still working with and without DMA. Note that the latter increases ROM usage by 1KB and RAM usage by18B.
`tests/periph_dma` is also working.

**Comment by bergzand** [comment]
> I just tested `tests/pkg_lvgl` on stm32f429disc1 and it's still working with and without DMA.

This should be fine here. While the ili9341 never reads the received SPI data (it can operate without MISO line), the RX DMA channel is still configured and must have a valid configuration for this to work.

**Comment by bergzand** [comment]
I can provide output for the nucleo-f401re and the nucleo-f446re boards here.

File: boards/msbiot/include/periph_conf.h [hunk 1]
```diff
@@ -27,6 +27,24 @@
 extern "C" {
 #endif
 
+/**
+ * @name    DMA streams configuration
+ * @{
+ */
+#ifdef MODULE_PERIPH_DMA
+static const dma_conf_t dma_config[] = {
+    { .stream = 11 },   /* DMA2 Stream 3 - SPI1_TX */
```
Comment (aabadie) [inline boards/msbiot/include/periph_conf.h:36]: How do you determine the value set to stream ? I can't find anything related to 11 or 10 in the reference manual. Shouldn't it be 3 here and 2 below ?

Comment (aabadie) [reply boards/msbiot/include/periph_conf.h:36]: Oh, I think I have it now: since it's DMA2, the 8 streams of DMA1 are added. So makes sense.

Comment (bergzand) [reply boards/msbiot/include/periph_conf.h:36]: Yes, it is not completely intuitive, we should probably add a `DMA_STREAM(dma_dev, stream)` macro at some point for clarity. It is described [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32_common/include/periph_cpu_common.h#L273-L297) and [here](https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32_common/periph/dma.c#L99-L107)

File: boards/nucleo-f401re/include/periph_conf.h [hunk 2]
```diff
@@ -28,6 +28,32 @@
...
+ * @{
+ */
+#ifdef MODULE_PERIPH_DMA
+static const dma_conf_t dma_config[] = {
+    { .stream = 11 },   /* DMA2 Stream 3 - SPI1_TX */
+    { .stream = 10 },   /* DMA2 Stream 2 - SPI1_RX */
+    { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
+    { .stream = 3 },    /* DMA1 Stream 3 - SPI2_RX */
+    { .stream = 5 },    /* DMA1 Stream 5 - SPI3_TX */
+    { .stream = 0 },    /* DMA1 Stream 0 - SPI2_RX */
```
Comment (aabadie) [inline boards/nucleo-f401re/include/periph_conf.h:42]:
```suggestion
    { .stream = 0 },    /* DMA1 Stream 0 - SPI3_RX */
```

File: boards/nucleo-f446re/include/periph_conf.h [hunk 3]
```diff
@@ -29,6 +29,32 @@
...
+ * @{
+ */
+#ifdef MODULE_PERIPH_DMA
+static const dma_conf_t dma_config[] = {
+    { .stream = 11 },   /* DMA2 Stream 3 - SPI1_TX */
+    { .stream = 10 },   /* DMA2 Stream 2 - SPI1_RX */
+    { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
+    { .stream = 3 },    /* DMA1 Stream 3 - SPI2_RX */
+    { .stream = 5 },    /* DMA1 Stream 5 - SPI3_TX */
+    { .stream = 0 },    /* DMA1 Stream 0 - SPI2_RX */
```
Comment (aabadie) [inline boards/nucleo-f446re/include/periph_conf.h:43]:
```suggestion
    { .stream = 0 },    /* DMA1 Stream 0 - SPI3_RX */
```

File: boards/stm32f4discovery/include/periph_conf.h [hunk 4]
```diff
@@ -29,6 +29,32 @@
...
+ * @{
+ */
+#ifdef MODULE_PERIPH_DMA
+static const dma_conf_t dma_config[] = {
+    { .stream = 11 },   /* DMA2 Stream 3 - SPI1_TX */
+    { .stream = 10 },   /* DMA2 Stream 2 - SPI1_RX */
+    { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
+    { .stream = 3 },    /* DMA1 Stream 3 - SPI2_RX */
+    { .stream = 5 },    /* DMA1 Stream 5 - SPI3_TX */
+    { .stream = 0 },    /* DMA1 Stream 0 - SPI2_RX */
```
Comment (aabadie) [inline boards/stm32f4discovery/include/periph_conf.h:43]:
```suggestion
    { .stream = 0 },    /* DMA1 Stream 0 - SPI3_RX */
```

File: boards/stm32f4discovery/include/periph_conf.h [hunk 5]
```diff
@@ -29,6 +29,32 @@
...
+ * @name    DMA streams configuration
+ * @{
+ */
+#ifdef MODULE_PERIPH_DMA
+static const dma_conf_t dma_config[] = {
+    { .stream = 11 },   /* DMA2 Stream 3 - SPI1_TX */
+    { .stream = 10 },   /* DMA2 Stream 2 - SPI1_RX */
+    { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
+    { .stream = 3 },    /* DMA1 Stream 3 - SPI2_RX */
+    { .stream = 5 },    /* DMA1 Stream 5 - SPI3_TX */
```
Comment (aabadie) [inline boards/stm32f4discovery/include/periph_conf.h:42]: SPI3 is not used on this board. Not sure we want to keep it ?

**Comment by aabadie** [comment]
I compared the changes of this PR against the reference manuals of the CPUs involved and all configurations are correct AFAICT. Just found some minor nits.

I would be OK to merge this PR without testing on all boards. Just the ones we have around is OK (stm32f429i-disc1, nucleo-f401re and nucleo-f446re).

File: boards/stm32f4discovery/include/periph_conf.h (hunk 5 above)
Comment (bergzand) [reply boards/stm32f4discovery/include/periph_conf.h]: I removed it, let's keep this clean

File: boards/nucleo-f401re/include/periph_conf.h (hunk 2 above)
Comment (bergzand) [reply boards/nucleo-f401re/include/periph_conf.h]: Fixed!

File: boards/nucleo-f446re/include/periph_conf.h (hunk 3 above)
Comment (bergzand) [reply boards/nucleo-f446re/include/periph_conf.h]: Fixed!

File: boards/stm32f4discovery/include/periph_conf.h (hunk 4 above)
Comment (bergzand) [reply boards/stm32f4discovery/include/periph_conf.h]: Fixed! (and then removed)

File: boards/stm32f4discovery/include/periph_conf.h [hunk 6]
```diff
@@ -29,6 +29,30 @@
...
+    { .stream = 10 },   /* DMA2 Stream 2 - SPI1_RX */
+    { .stream = 4 },    /* DMA1 Stream 4 - SPI2_TX */
+    { .stream = 3 },    /* DMA1 Stream 3 - SPI2_RX */
+};
+
+#define DMA_0_ISR           isr_dma2_stream3
+#define DMA_1_ISR           isr_dma2_stream2
+#define DMA_2_ISR           isr_dma1_stream4
+#define DMA_3_ISR           isr_dma1_stream3
+#define DMA_4_ISR           isr_dma1_stream5
```
Comment (aabadie) [inline boards/stm32f4discovery/include/periph_conf.h:48]: The 2 last ISR defines have to be removed as well ;)

Comment (bergzand) [reply boards/stm32f4discovery/include/periph_conf.h]: Arrrgh, thanks, removed.

**Comment by bergzand** [comment]
<details>
  <summary>Nucleo-f401re</summary>

  tests/periph_dma:
  ```
  2020-05-25 11:30:24,941 # Help: Press s to start test, r to print it is ready
  s
  2020-05-25 11:30:26,802 # START
  2020-05-25 11:30:26,811 # main(): This is RIOT! (Version: 2020.07-devel-687-g201d6-pr/boards_stm32f4/add_spi_dma)
  2020-05-25 11:30:26,814 # DMA is working
  ```

  tests/periph_spi:
  ```
  init 0 0 4 -1 0
  2020-05-25 11:36:10,469 # init 0 0 4 -1 0
  2020-05-25 11:36:10,476 # SPI_DEV(0) initialized: mode: 0, clk: 4, cs_port: -1, cs_pin: 0
  > bench
  2020-05-25 11:36:11,662 #  bench
  2020-05-25 11:36:11,668 # ### Running some benchmarks, all values in [us] ###
  2020-05-25 11:36:11,671 # ### Test                              Transfer time   user time
  2020-05-25 11:36:11,671 #
  2020-05-25 11:36:11,692 #  1 - write 1000 times 1 byte:                 17359   17361
  2020-05-25 11:36:11,722 #  2 - write 1000 times 2 byte:                 20819   20821
  2020-05-25 11:36:11,813 #  3 - write 1000 times 100 byte:               96096   19193
  2020-05-25 11:36:11,855 #  4 - write 1000 times 1 byte to register:     34347   34349
  2020-05-25 11:36:11,896 #  5 - write 1000 times 2 byte to register:     37870   37873
  2020-05-25 11:36:12,017 #  6 - write 1000 times 100 byte to register:   113096  36194
  2020-05-25 11:36:12,046 #  7 - read 1000 times 2 byte:                  20855   20857
  2020-05-25 11:36:12,142 #  8 - read 1000 times 100 byte:                96096   19193
  2020-05-25 11:36:12,185 #  9 - read 1000 times 2 byte from register:    37882   37885
  2020-05-25 11:36:12,301 # 10 - read 1000 times 100 byte from register:  113168  36170
  2020-05-25 11:36:12,328 # 11 - transfer 1000 times 2 byte:              20903   20905
  2020-05-25 11:36:12,430 # 12 - transfer 1000 times 100 byte:            96168   19170
  2020-05-25 11:36:12,469 # 13 - transfer 1000 times 2 byte to register:  37953   37956
  2020-05-25 11:36:12,588 # 14 - transfer 1000 times 100 byte to register:113192  36290
  2020-05-25 11:36:13,589 # -- - SUM:                                     855804  394217
  2020-05-25 11:36:13,590 #
  2020-05-25 11:36:13,594 # ### All runs complete ###
  ```

</details>

<details>
<summary>nucleo-f446re</summary>

tests/periph_dma:
```
2020-05-25 11:40:28,474 # START
2020-05-25 11:40:28,482 # main(): This is RIOT! (Version: 2020.07-devel-687-g201d6-pr/boards_stm32f4/add_spi_dma)
2020-05-25 11:40:28,483 # DMA is working
```

tests/periph_spi:
```
init 0 0 4 -1 0
2020-05-25 11:39:24,889 # init 0 0 4 -1 0
2020-05-25 11:39:24,894 # SPI_DEV(0) initialized: mode: 0, clk: 4, cs_port: -1, cs_pin: 0
bench
2020-05-25 11:39:26,890 #  bench
2020-05-25 11:39:26,895 # ### Running some benchmarks, all values in [us] ###
2020-05-25 11:39:26,898 # ### Test                              Transfer time   user time
2020-05-25 11:39:26,898 # 
2020-05-25 11:39:26,912 #  1 - write 1000 times 1 byte:                 9912    9913
2020-05-25 11:39:26,928 #  2 - write 1000 times 2 byte:                 12223   12225
2020-05-25 11:39:27,014 #  3 - write 1000 times 100 byte:               82823   11670
2020-05-25 11:39:27,039 #  4 - write 1000 times 1 byte to register:     19657   19658
2020-05-25 11:39:27,066 #  5 - write 1000 times 2 byte to register:     22001   22003
2020-05-25 11:39:27,163 #  6 - write 1000 times 100 byte to register:   92645   21491
2020-05-25 11:39:27,180 #  7 - read 1000 times 2 byte:                  12190   12191
2020-05-25 11:39:27,266 #  8 - read 1000 times 100 byte:                82845   11690
2020-05-25 11:39:27,293 #  9 - read 1000 times 2 byte from register:    21989   21991
2020-05-25 11:39:27,390 # 10 - read 1000 times 100 byte from register:  92645   21470
2020-05-25 11:39:27,407 # 11 - transfer 1000 times 2 byte:              12256   12258
2020-05-25 11:39:27,494 # 12 - transfer 1000 times 100 byte:            82911   11757
2020-05-25 11:39:27,521 # 13 - transfer 1000 times 2 byte to register:  22046   22047
2020-05-25 11:39:27,618 # 14 - transfer 1000 times 100 byte to register:92711   21535
2020-05-25 11:39:28,621 # -- - SUM:                                     658854  231899
2020-05-25 11:39:28,622 # 
2020-05-25 11:39:28,623 # ### All runs complete ###
```
</details>

**Comment by bergzand** [comment]
Squashed!

**Comment by bergzand** [comment]
Thanks!

**Pull request merged**